### PR TITLE
Create generic einops.einop operation to do any of rearrange/reduce/repeat

### PR DESCRIPTION
**Pull request by MilesCranmer**
As mentioned in #84, the einop operation can be inferred based on pattern alone. This PR adds a function to do this: `einops.einop`. For example:

Reduction:

```python
import numpy as np
from einops import einop

x = np.random.randn(100, 5, 3)

einop(x, 'i j k -> i j', reduction='sum').shape
>>> (100, 5)
```

Rearrange:
```python
einop(x, 'i j k -> k i j').shape
>>> (3, 100, 5)
```

Repeat:
```python
einop(x, 'i j k -> i j k l', l=10).shape
(100, 5, 3, 10)
```

There are also assertions to catch bugs as mentioned in #84. I also added some tests for these operations: test12 through test14.

Interested to hear what you think.
Cheers!
Miles

### CODE REVIEW

File: einops/einops.py [hunk 1]
```diff
@@ -323,6 +323,47 @@ def _prepare_transformation_recipe(pattern: str,
...
+
+    return op
+
+def einop(tensor, pattern: str, reduction=None, **axes_lengths: int):
+    """Perform either reduce, rearrange, or repeat depending on pattern"""
+    op = _match_einop(pattern, reduction, **axes_lengths)
+
+    if op == 'rearrange':
+        if reduction is not None:
+            raise EinopsError('Do not pass reduction for rearrange pattern: {}'.format(pattern))
```
Comment (cgarciae) [inline einops/einops.py:356]: I think this error message is "leaking" implementation details (the use of rearrange). Consider something like:

> Got reduction operation but there is no dimension to reduce in pattern: '{pattern}'

File: einops/einops.py [hunk 2]
```diff
@@ -323,6 +323,47 @@ def _prepare_transformation_recipe(pattern: str,
...
+        return rearrange(tensor, pattern, **axes_lengths)
+    elif op == 'reduce':
+        if reduction is None:
+            raise EinopsError('Missing reduction operation for reduce pattern: {}'.format(pattern))
```
Comment (cgarciae) [inline einops/einops.py:360]: Same

**Comment by genghisun** [comment]
Such a great feature! 👍 Any update on this?

**Comment by MilesCranmer** [comment]
^See #84 for discussion. Currently, the maintainer is against adding this

File: einops/einops.py (hunk 1 above)
Comment (MilesCranmer) [reply einops/einops.py]: Thanks, fixed.

**Comment by cgarciae** [comment]
I've been waiting for this PR to land for a while, in the meantime I've extracted this into this [einop](https://github.com/cgarciae/einop) package.

**Comment by FrancescoSaverioZuppichini** [comment]
guys?

**Comment by MilesCranmer** [comment]
I wish we could, but nothing we can do unfortunately. If you have a compelling argument to convince the maintainer, you can add it to #84. In the meantime you can use @cgarciae’s package

**Comment by UFO-101** [comment]
@MilesCranmer Thoughts on doing a similar thing to einsum? https://github.com/arogozhnikov/einops/issues/84#issuecomment-1641109828?

**Pull request closed without merging**